src/bitExpert/Disco/Proxy/Configuration/MethodGenerator/BeanMethod.php

```diff
@@ -120,7 +128,15 @@ public static function generateMethod(
             }
 
             $body .= $innerpadding . '$instance = parent::' . $methodName . '(' . $methodParamTpl . ');' . "\n";
-            $body .= $innerpadding . '$this->initializeBean($instance, "' . $methodName . '");' . "\n";
+            $body .= $innerpadding . '$this->initializeBean($instance, "' . $methodName . '");' . "\n\n";
+            $body .= $innerpadding . 'if (!is_a($instance, \'' . $beanType . '\')) {' . "\n";
```

Comment:
Can we replace the is_a() function call with an instanceof check? That might potentially be faster as it is not a method call.

In addition the check should be moved before the $this->initializeBean(). If the expected value is not returned, there is no need to run the initializer logic. Failing early makes sense to me in this case.